File: 101-symmetric-tree/101-symmetric-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if( !root )
            return true;
        
        
        std::queue<TreeNode *> qL;
        std::queue<TreeNode *> qR;
        qL.push(root);
        qR.push(root);
        
        while (!qL.empty()) {
            int qSize = qL.size();
            
            for( int i = 0; i < qSize; i++) {
                TreeNode *nodeL = qL.front();
                qL.pop();
                TreeNode *nodeR = qR.front();
                qR.pop();
                
                if( nodeL->left && nodeR->right ) {
                    qL.push(nodeL->left);
                    qR.push(nodeR->right);
                }
                else if (nodeL->left || nodeR->right)
                    return false;
                
                if (nodeL->right && nodeR->left ) {
                    qL.push(nodeL->right);
                    qR.push(nodeR->left);
                }
                else if (nodeL->right || nodeR->left)
                    return false;
                
                if (nodeL->val != nodeR->val)
                    return false;
            }
        }
        
        return true;
    }
};

File: 101-symmetric-tree/101-symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root is None:
            return True
        
        q = collections.deque([(root, root)])
        while len(q) != 0:
            numNodes = len(q)
            
            for _ in range(numNodes):
                nodeL,nodeR = q.popleft()
                
                if nodeL.left is not None and nodeR.right is not None:
                    q.append((nodeL.left, nodeR.right))
                elif nodeL.left is not None or nodeR.right is not None:    
                        return False

                if nodeL.right is not None and nodeR.left is not None:
                    q.append((nodeL.right, nodeR.left))
                elif nodeL.right is not None or nodeR.left is not None:    
                        return False
                
                if nodeL.val != nodeR.val:
                    return False

        return True
    
    
        